lib/semantics/check-do-stmt.cc

```diff
@@ -406,31 +410,50 @@ class DoStmtContext {
       // TODO: Mark the following message as a warning when we have warnings
       context_.Say(sourceLocation, "DO controls should be INTEGER"_en_US);
     } else {
-      context_.Say(sourceLocation, "DO controls should be INTEGER"_err_en_US);
+      SayBadDoControl(sourceLocation);
     }
   }
 
   void CheckDoVariable(const parser::ScalarName &scalarName) {
-    const DeclTypeSpec *symType{scalarName.thing.symbol->GetType()};
-    if (symType->IsNumeric(TypeCategory::Integer)) {
-      return;  // No warnings or errors for INTEGER
-    }
     const parser::CharBlock &sourceLocation{scalarName.thing.source};
-    CheckDoControl(sourceLocation, symType->IsNumeric(TypeCategory::Real));
+    const Symbol *symbol{scalarName.thing.symbol};
+    if (!symbol) {
+      SayBadDoControl(sourceLocation);
```

Comment:
Do you have a test that covers this case? I would expect that name resolution would have already produced an error about this name so you can just ignore it here.

Reply:
dosemantics03.f90 has a test for this.  The problem happens when "IMPLICIT NONE" is in effect, and
you use an undeclared variable as the DO variable.  The failing test starts at line 150.